Signal/src/ViewControllers/MenuActionsViewController.swift

```diff
@@ -0,0 +1,402 @@
+                        newFocusFrame.origin.y = oldFocusFrame.origin.y - overlap
+
+                        snapshotView.frame = newFocusFrame
+
+                        let offset = -overlap
+                        self.presentationFocusOffset = offset
+                        self.delegate?.menuActions(self, isPresentingWithVerticalFocusChange: offset)
+        },
+                       completion: nil)
+    }
+
+    private func animateDismiss(action: MenuAction?) {
+        guard let actionSheetViewVerticalConstraint = self.actionSheetViewVerticalConstraint else {
+            owsFail("\(self.logTag) in \(#function) actionSheetVerticalConstraint was unexpectedly nil")
+            self.delegate?.menuActionsDidHide(self)
+            return
+        }
+
+        guard let snapshotView = self.snapshotView else {
+            owsFail("\(self.logTag) in \(#function) snapshotView was unexpectedly nil")
+            self.delegate?.menuActionsDidHide(self)
+            return
+        }
+
+        guard let presentationFocusOffset = self.presentationFocusOffset else {
+            owsFail("\(self.logTag) in \(#function) presentationFocusOffset was unexpectedly nil")
+            self.delegate?.menuActionsDidHide(self)
+            return
+        }
+
+        self.actionSheetView.superview?.layoutIfNeeded()
+        NSLayoutConstraint.deactivate([actionSheetViewVerticalConstraint])
+
+        let dismissDuration: TimeInterval = 0.2
+        self.actionSheetViewVerticalConstraint = self.actionSheetView.autoPinEdge(.top, to: .bottom, of: self.view)
+        UIView.animate(withDuration: dismissDuration,
+                       delay: 0,
+                       options: .curveEaseOut,
+                       animations: {
+                        self.view.backgroundColor = UIColor.clear
+                        self.actionSheetView.superview?.layoutIfNeeded()
+                        snapshotView.frame.origin.y -= presentationFocusOffset
+                        // this helps when focused view is above navbars, etc.
+                        snapshotView.alpha = 0
+                        self.ensureDelegateIsInformedOfDismissalAnimation()
+        },
+                       completion: { _ in
+                        self.view.isHidden = true
+                        self.ensureDelegateIsInformedThatDisappearenceCompleted()
+                        if let action = action {
+                            action.block(action)
+                        }
+        })
+    }
+
+    var didInformDelegateThatDisappearenceCompleted = false
+    func ensureDelegateIsInformedThatDisappearenceCompleted() {
+        guard !didInformDelegateThatDisappearenceCompleted else {
+            Logger.debug("\(logTag) in \(#function) ignoring redundant 'disappeared' notification")
+            return
+        }
+        didInformDelegateThatDisappearenceCompleted = true
+
+        self.delegate?.menuActionsDidHide(self)
```

Comment:
> Sorry, I meant to ask if isDismissingWithVerticalFocusChange could be called twice?

The new method: `ensureDelegateIsInformedOfDismissalAnimation` exists precisely to ensure that `isDismissingWithVerticalFocusChange` is only called once, (hence the `guard !didInformDelegateOfDismissalAnimation`)

Since, as you correctly pointed out, there are multiple ways to exit the MenuActions view (e.g. the user could explicitly dismiss it, or it could get implicitly dismissed by backgrounding the app) we can sprinkle around the `ensure...` methods at all possible exit points, and be confident the delegate will never be informed more than once.